File: src/java/graphs/200.number-of-islands.java
/*
 * @lc app=leetcode id=200 lang=java
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length ==0) {
            return 0;
        }
        int numberOfOIslands  =0;
        for(int i=0; i<grid.length;i++){

            for (int j=0; j< grid[i].length;j++) {
                if(grid[i][j] =='1'){
                    dfs(grid,i,j);
                    numberOfOIslands ++;
                }
            }
        }

        return numberOfOIslands;
        }


        public void dfs(char[][] grid, int i, int j){

            if(i < 0 || i>= grid.length || j <0 || j >= grid[i].length
            || grid[i][j] == '0') {
                return;
            }

grid[i][j] ='0';
            //let us dfs of all 4 directions


            dfs(grid,i-1,j);
            dfs(grid,i,j+1);
            dfs(grid,i+1,j);
            dfs(grid,i,j-1);


        }
    }
// @lc code=end

File: src/go/graphs/200.number-of-islands.go
package main
package main

/*
 * @lc app=leetcode id=200 lang=golang
 *
 * [200] Number of Islands
 */

// @lc code=start
func numIslands(grid [][]byte) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}
	
	rows, cols := len(grid), len(grid[0])
	count := 0
	
	var dfs func(r, c int)
	dfs = func(r, c int) {
		if r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] != '1' {
			return
		}
		grid[r][c] = '0' // Mark as visited
		dfs(r+1, c)
		dfs(r-1, c)
		dfs(r, c+1)
		dfs(r, c-1)
	}
	
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == '1' {
				count++
				dfs(r, c)
			}
		}
	}
	
	return count
}
// @lc code=end